src/build/mod.rs

```diff
@@ -547,12 +545,56 @@ impl Internals {
         };
         trace!("Specified work: {:?}", work);
 
-        match work {
+        let cargo_build = if let WorkStatus::NeedsCargo(..) = work {
+            true
+        } else {
+            false
+        };
+
+        let result = match work {
             // Cargo performs the full build and returns
             // appropriate diagnostics/analysis data
             WorkStatus::NeedsCargo(package_arg) => cargo::cargo(self, package_arg, progress_sender),
             WorkStatus::Execute(job_queue) => job_queue.execute(self, progress_sender),
+        };
+
+        if let BuildResult::Success(.., true) = result {
+            let elapsed = start.elapsed();
+            *self.last_build_duration.write().unwrap() = Some(elapsed);
+            info!(
+                "{} finished in {:.1?}",
+                if cargo_build { "cargo" } else { "rustc" },
+                elapsed
+            );
         }
+
+        result
+    }
+
+    /// Returns a pre-build wait time facilitating build debouncing.
+    ///
+    /// Uses client configured value, or attempts to infer an appropriate
+    /// duration.
+    fn build_wait(&self) -> Duration {
+        self.config
+            .lock()
+            .unwrap()
+            .wait_to_build
+            .map(Duration::from_millis)
+            .unwrap_or_else(|| match *self.last_build_duration.read().unwrap() {
+                None => Duration::from_millis(1500),
+                Some(duration) => {
+                    if duration < Duration::from_millis(300) {
+                        Duration::from_millis(0)
+                    } else if duration < Duration::from_secs(1) {
+                        Duration::from_millis(200)
+                    } else if duration < Duration::from_secs(5) {
+                        Duration::from_millis(500)
+                    } else {
+                        Duration::from_millis(1500)
```

Comment:
Could you please unify the `None` and default case? I think it'd dedup a bit and be more readable. Something like:
```rust
	Some(duration) if duration < Duration::from_secs(5) {
		if duration < Duration::from_millis(300) {
			Duration::from_millis(0)
		} else if duration < Duration::from_secs(1) {
			Duration::from_millis(200)
		} else {
			Duration::from_millis(500)
		}
	},
	_ => Duration::from_millis(1500),
```

Reply:
Good idea!